cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,685 @@
+                   return get_column_metadata(row_group_index, src_index, schema_idx).type;
+                 });
+
+  return parquet_types;
+}
+
+std::optional<std::vector<std::vector<size_type>>> aggregate_reader_metadata::apply_bloom_filters(
+  host_span<std::unique_ptr<datasource> const> sources,
+  host_span<std::vector<size_type> const> input_row_group_indices,
+  host_span<data_type const> output_dtypes,
+  host_span<int const> output_column_schemas,
+  std::reference_wrapper<ast::expression const> filter,
+  rmm::cuda_stream_view stream) const
+{
+  // Number of input table columns
+  auto const num_input_columns = static_cast<cudf::size_type>(output_dtypes.size());
+
+  auto mr = cudf::get_current_device_resource_ref();
+
+  // Total number of row groups to process.
+  auto const num_row_groups = std::accumulate(
+    input_row_group_indices.begin(),
+    input_row_group_indices.end(),
+    size_t{0},
+    [](size_t sum, auto const& per_file_row_groups) { return sum + per_file_row_groups.size(); });
+
+  // Collect equality literals for each input table column
+  auto const equality_literals =
+    equality_literals_collector{filter.get(), num_input_columns}.get_equality_literals();
+
+  // Collect schema indices of columns with equality predicate(s)
+  std::vector<cudf::size_type> equality_col_schemas;
+  std::for_each(thrust::make_counting_iterator<size_t>(0),
+                thrust::make_counting_iterator(output_column_schemas.size()),
+                [&](auto col_idx) {
+                  // Only for columns that have a non-empty list of literals associated with it
+                  if (equality_literals[col_idx].size()) {
+                    equality_col_schemas.emplace_back(output_column_schemas[col_idx]);
+                  }
+                });
+
+  // Return early if no column with equality predicate(s)
+  if (equality_col_schemas.empty()) { return std::nullopt; }
+
+  // Read a vector of bloom filter bitset device buffers for all column with equality
+  // predicate(s) across all row groups
+  auto const bloom_filter_data = read_bloom_filters(
+    sources, input_row_group_indices, equality_col_schemas, num_row_groups, stream);
+
+  // No bloom filter buffers, return the original row group indices
+  if (bloom_filter_data.empty()) { return std::nullopt; }
+
+  // Get parquet types for the predicate columns
+  auto const parquet_types = get_parquet_types(input_row_group_indices, equality_col_schemas);
+
+  // Copy bloom filter bitset buffer pointers and sizes to device
+  std::vector<void*> h_buffer_ptrs(bloom_filter_data.size());
+  std::vector<size_t> h_buffer_sizes(bloom_filter_data.size());
+  std::for_each(thrust::make_counting_iterator<size_t>(0),
+                thrust::make_counting_iterator<size_t>(bloom_filter_data.size()),
+                [&](auto i) {
+                  auto const& buffer = bloom_filter_data[i];
+                  // Bitset ptr must be non-const to be used in cuco::bloom_filter.
```

Comment:
yuck
Is this something cuco can change?

Reply:
I don’t see any reason to prevent the use of a const pointer here. The content pointed by the pointer must be mutable, but the pointer itself can be const.

Reply:
Oh I see, the issue is that `bloom_filter_data` should not be const

Reply:
If cuco is actually changing the content, we should make `bloom_filter_data` and `buffer` non-const. Then the `const_cast` won't be needed.